aea/decision_maker/messages/transaction.py

```diff
@@ -95,60 +95,126 @@ def __init__(self, performative: Union[str, Performative],
+            assert isinstance(self.transaction_id, str), "Transaction_id must of type str."
+            assert isinstance(self.sender, Address), "Sender must be of type address."
+            assert isinstance(self.counterparty, Address), "Counterparty must be of type address"
+            assert self.sender != self.counterparty, "Sender must be different of counterparty."
+            assert isinstance(self.is_sender_buyer, bool), "Is_sender_buyer must be of type bool."
+            assert isinstance(self.currency_pbk, str), "Currency_pbk must be of type str."
+            assert isinstance(self.amount, int), "Amount must be of type int"
+            assert self.amount >= 0, "Amount must be more than zero."
+            assert isinstance(self.sender_tx_fee, int), "Sender_tx_fee must be of type int."
+            assert self.sender_tx_fee >= 0, "Sender transaction fee must be greater or equal to zero."
+            assert isinstance(self.counterparty_tx_fee, int), "Counter_tx_fee must be of type int."
+            assert self.counterparty_tx_fee >= 0, "Counterparty transaction fee must be greater or equal to zero."
+            assert isinstance(self.ledger_id, str) and self.ledger_id in SUPPORTED_LEDGER_IDS, "Ledger_id must be str and " \
+                                                                                               "must in the supported ledger ids."
+
+            if self.info is not None:
+                assert isinstance(self.info, Dict)
+                for key, value in self.info.items():
                     assert type(key) == str
-            assert self.is_set("quantities_by_good_pbk")
-            quantities_by_good_pbk = self.get("quantities_by_good_pbk")
-            if quantities_by_good_pbk is not None:
-                assert type(quantities_by_good_pbk) == dict
-                for key, value in quantities_by_good_pbk.items():
+            if self.quantities_by_good_pbk is not None:
+                assert type(self.quantities_by_good_pbk) == dict
+                for key, value in self.quantities_by_good_pbk.items():
                     assert type(key) == str and type(value) == int
-                quantities_by_good_pbk = cast(Dict[str, int], quantities_by_good_pbk)
-                assert len(quantities_by_good_pbk.keys()) == len(set(quantities_by_good_pbk.keys()))
-                assert all(quantity >= 0 for quantity in quantities_by_good_pbk.values())
-            assert self.is_set("transaction_digest")
-            transaction_digest = self.get("transaction_digest")
-            if transaction_digest is not None:
-                assert type(transaction_digest) == str
+                assert len(self.quantities_by_good_pbk.keys()) == len(set(self.quantities_by_good_pbk.keys()))
+                assert all(quantity >= 0 for quantity in self.quantities_by_good_pbk.values())
+            if self.transaction_digest is not None:
```

Comment:
need to do `self._transaction_digest` otherwise this will throw if None!

Reply:
There is not variable self._transaction_digest. You are passing the transaction_digest through the contractor but you don't save it in a variable _transaction_digest.

Reply:
`assert self.is_set("transaction_digest")
  transaction_digest = self.get("transaction_digest")		
  if transaction_digest is not None:`

this is the previous implementation. Again if you don't set the transaction_digest you are going to raise an Assertion error.